modules/images/webp-uploads/load.php

```diff
@@ -683,3 +683,190 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
 	return rest_ensure_response( $data );
 }
 add_filter( 'rest_prepare_attachment', 'webp_uploads_update_rest_attachment', 10, 3 );
+
+/**
+ * Inspect if the current call to `wp_update_attachment_metadata()` was done from within the context
+ * of an edit to an attachment either restore or other type of edit, in that case we perform operations
+ * to save the sources properties, specifically for the `full` size image due this is a virtual image size.
+ *
+ * @since n.e.x.t
+ *
+ * @see wp_update_attachment_metadata()
+ *
+ * @param array $data The current metadata of the attachment.
+ * @param int   $attachment_id The ID of the current attachment.
+ * @return array The updated metadata for the attachment to be stored in the meta table.
+ */
+function webp_wp_update_attachment_metadata( $data, $attachment_id ) {
+
+	$trace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 10 );
+
+	foreach ( $trace as $element ) {
+		if ( ! isset( $element['function'] ) ) {
+			continue;
+		}
+
+		switch ( $element['function'] ) {
+			case 'wp_save_image':
+				// Right after an image has been edited.
+				return webp_uploads_backup_sources( $attachment_id, $data );
+			case 'wp_restore_image':
+				// When an image has been restored.
+				return webp_uploads_restore_image( $attachment_id, $data );
+		}
+	}
+
+	return $data;
+}
+
+add_filter( 'wp_update_attachment_metadata', 'webp_wp_update_attachment_metadata', 10, 2 );
+
+/**
+ * Before saving the metadata of the image store a backup values for the sources and file property
+ * those files would be used and deleted by the backup mechanism, right after the metadata has
+ * been updated. It removes the current sources property due once this function is executed
+ * right after an edit has taken place and the current sources are no longer accurate.
+ *
+ * @since n.e.x.t
+ *
+ * @param int   $attachment_id The ID representing the attachment.
+ * @param array $data The current metadata of the attachment.
+ * @return array The updated metadata for the attachment.
+ */
+function webp_uploads_backup_sources( $attachment_id, $data ) {
+	if ( ! isset( $data['sources'] ) ) {
+		return $data;
+	}
+
+	$data['_sources'] = $data['sources'];
+	// Remove the current sources as at this point the current values are no longer accurate.
+	unset( $data['sources'] );
```

Comment:
This is a bit problematic in terms of core, since core supports multiple revisions for edited images, which this code here doesn't. How about we follow the WP core implementation from `wp_save_image()` and we use a new `_wp_attachment_backup_sources` meta key to store the values? To elaborate a bit further:
1. Get the current value of `_wp_attachment_backup_sources` into `$backup_sources`. Fall back to an empty array if not set.
2. Define a suffix variable similar like in core, e.g. `$suffix = time() . rand( 100, 999 )`.
3. If `$backup_sources['sources-orig']` already exists, set the current `sources` in `$backup_sources["sources-$suffix"]`. Otherwise (no backup exists yet), use `$backup_sources['sources-orig']`.
4. Unset `sources` from the original metadata entirely (since we don't yet support generating sources for image edits, which will be part of #158).
5. Set `_wp_attachment_backup_sources` with the updated `$backup_sources`.

Doesn't have to be exactly like the above, but we need to support history, and I think a separate meta key makes sense here to follow core.

Reply:
> This is a bit problematic in terms of core, since core supports multiple revisions for edited images, which this code here doesn't.

Agree, however, this is already covered by this conditional.

- https://github.com/WordPress/performance/blob/4820501b979d4e710baed19bd75d4597b628562b/modules/images/webp-uploads/load.php#L838

> Get the current value of _wp_attachment_backup_sources into $backup_sources. Fall back to an empty array if not set.

I would argue that we should fallback to not set the `sources` attributes if we don't have any data to process rather than an empty array, however I do see the benefit for consistency, so let me do that.

> Define a suffix variable similar like in core, e.g. $suffix = time() . rand( 100, 999 ).

I don't think we should do that, instead we should apply this logic instead to determine the next available slot for full size images.

https://github.com/WordPress/performance/blob/4820501b979d4e710baed19bd75d4597b628562b/modules/images/webp-uploads/load.php#L842-L855

The main reason is that would give us the next available slot for full size images instead, if we create a custom hash we are not going to be able to determine which sources belong to which image sizes as if you perform multiple images multiple edited versions exists so in the case you performed and saved 5 edited operations and `IMAGE_EDIT_OVERWRITE` is not defined you would end up with 5 different `full-{hash or orig for the fist time}` so creating a custom hash from our end would not be beneficial and prevent us from relate the correct full versions of the edited images instead we should use the hash already stored on the backup image sizes.

> If $backup_sources['sources-orig'] already exists, set the current sources in $backup_sources["sources-$suffix"]. Otherwise (no backup exists yet), use $backup_sources['sources-orig'].

This logic is covered and answered on the previous item.

> Unset sources from the original metadata entirely (since we don't yet support generating sources for image edits, which will be part of https://github.com/WordPress/performance/issues/158).
Set _wp_attachment_backup_sources with the updated $backup_sources.

Right this is already covered here: https://github.com/WordPress/performance/blob/4820501b979d4e710baed19bd75d4597b628562b/modules/images/webp-uploads/load.php#L743

Reply:
> I would argue that we should fallback to not set the `sources` attributes if we don't have any data to process rather than an empty array, however I do see the benefit for consistency, so let me do that.

The backup sources would be what's already backed up. If they're not set/empty, that just means there is no backup yet, but we still have to do something on that data, populate it the new backup sources.

> I don't think we should do that, instead we should apply this logic instead to determine the next available slot for full size images.

Good point, that makes sense to me.